crates/wasmprinter/src/lib.rs

```diff
@@ -2649,13 +2704,49 @@ impl Printer {
     fn print_component_imports(
         &mut self,
         state: &mut State,
-        types: &[TypeDef],
+        types: &[Type],
         parser: ComponentImportSectionReader,
     ) -> Result<()> {
         for import in parser {
             let import = import?;
             self.newline();
-            self.print_component_import(state, types, &import)?;
+            self.print_component_import(state, &import, true)?;
+            match import.ty {
+                ComponentTypeRef::Module(_) => {
+                    state.core.modules += 1;
+                }
+                ComponentTypeRef::Func(idx) => {
+                    let ty = state.ty(idx)?;
+                    match &types[ty] {
+                        Type::ComponentFunc(_) => {
+                            state.component.funcs.push(ty);
+                        }
+                        _ => bail!("index is not a function type"),
```

Comment:
I need to revisit this as this might be a holdover for when the binary encoding didn't provide alias type information so this state was maintained to compensate; we may be able to remove this entirely now.

Reply:
Now I remember why we have this convoluted type maintenance for printing components: start functions. The start function may produce a value, but to know if a value is being produced by the start function, we need to know the function's result type (unit produces no value, otherwise the value index space is incremented). Unfortunately, due to aliasing, we therefore have to track component and instance type information in case the function or its type is coming in from an alias.

The type tracking and validation is minimal and only enough to get component function types resolved.

Reply:
Ok well this probably isn't the PR to figure this out, so let's just eat all this and figure it out upstream.